tests/unit/test_space.py

```diff
@@ -438,3 +467,224 @@ def test_box_deepcopy() -> None:
+    "context_space, decision_space, num_context_features, num_decision_features",
+    [
+        (
+            DiscreteSearchSpace(tf.constant([[-0.5, 0.5]])),
+            DiscreteSearchSpace(tf.constant([[-0.5, -0.3], [1.2, 0.4]])),
+            2,
+            2,
+        ),
+        (DiscreteSearchSpace(tf.constant([[-0.5]])), Box([-1], [2]), 1, 1),
+        (Box([-1, -2], [2, 3]), DiscreteSearchSpace(tf.constant([[-0.5]])), 2, 1),
+        (Box([-1], [2]), Box([-1, -2], [2, 3]), 1, 2),
+    ],
+)
+def test_decomposable_space_can_get_context_component(
+    context_space: SearchSpace,
+    decision_space: SearchSpace,
+    num_context_features: int,
+    num_decision_features: int,
+) -> None:
+    space = DecomposableSearchSpace(context_space=context_space, decision_space=decision_space)
+    points = tf.random.uniform([10, num_context_features + num_decision_features])
+    npt.assert_array_equal(space.get_context_component(points), points[:, :num_context_features])
+
+
+@pytest.mark.parametrize(
+    "context_space, decision_space, num_context_features, num_decision_features",
+    [
+        (
+            DiscreteSearchSpace(tf.constant([[-0.5, 0.5]])),
+            DiscreteSearchSpace(tf.constant([[-0.5, -0.3], [1.2, 0.4]])),
+            2,
+            2,
+        ),
+        (DiscreteSearchSpace(tf.constant([[-0.5]])), Box([-1], [2]), 1, 1),
+        (Box([-1, -2], [2, 3]), DiscreteSearchSpace(tf.constant([[-0.5]])), 2, 1),
+        (Box([-1], [2]), Box([-1, -2], [2, 3]), 1, 2),
+    ],
+)
+def test_decomposable_space_can_get_decision_component(
+    context_space: SearchSpace,
+    decision_space: SearchSpace,
+    num_context_features: int,
+    num_decision_features: int,
+) -> None:
+    space = DecomposableSearchSpace(context_space=context_space, decision_space=decision_space)
+    points = tf.random.uniform([10, num_context_features + num_decision_features])
+    npt.assert_array_equal(space.get_decision_component(points), points[:, -num_decision_features:])
+
+
+@pytest.mark.parametrize(
+    "point",
+    [
+        tf.constant([-1.0, 0.0, -0.5, 0.5], dtype=tf.float64),
+        tf.constant([2.0, 3.0, -0.5, 0.5], dtype=tf.float64),
+    ],
+)
+def test_decomposable_space_contains_point(point: tf.Tensor) -> None:
+    context_space = Box([-1.0, -2.0], [2.0, 3.0])
+    decision_space = DiscreteSearchSpace(tf.constant([[-0.5, 0.5]], dtype=tf.float64))
+    decomposable_space = DecomposableSearchSpace(
+        context_space=context_space, decision_space=decision_space
+    )
+    assert point in decomposable_space
+
+
+@pytest.mark.parametrize(
+    "point",
+    [
+        tf.constant([-1.1, 0.0, -0.5, 0.5], dtype=tf.float64),  # just outside context space
+        tf.constant([-10, 10.0, -0.5, 0.5], dtype=tf.float64),  # well outside context space
+        tf.constant([2.0, 3.0, 2.0, 7.0], dtype=tf.float64),  # outside decision space
+        tf.constant([-10.0, -10.0, -10.0, -10.0], dtype=tf.float64),  # outside both
```

Comment:
Fun test case - wrong order of context and decision components

Reply:
So that's a "no"?

Reply:
done